# This notebook is for Coursera Capstone Project

Project :
          To obtain the zipcode information from the Wikipedia page on list of zip codes of Toronto,<br>
          convert it to pandas dataframe,<br>
          obtain latitude,longitude information for each zip code,<br>
          explore the nearby venues in each neighborhood using Foursquare,<br>
          perform Kmeans clustering on the neighborhoods based on their most common venues nearby<br>
          and finally show these labels on the Toronto Map using Folium.

Import the necessary libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

The data for the project is available on the Wikipedia page : https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M 
I will use pandas to read the table and convert it into a dataframe

In [2]:
# read the webpage using pandas_html method and get the first table in the page converted to a dataframe
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
dfs = pd.read_html(url)
Toronto_zipcodes=dfs[0]


In [3]:
Toronto_zipcodes.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
Toronto_zipcodes.shape

(180, 3)

Remove the rows for which Borough is not assigned

In [5]:
Toronto_zipcodes = Toronto_zipcodes[Toronto_zipcodes['Borough']!='Not assigned'].reset_index(drop=True)

In [6]:
Toronto_zipcodes.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [7]:
Toronto_zipcodes.shape

(103, 3)

In [9]:
Toronto_zipcodes['Postal Code'].nunique()

103

As the no. of rows is the same as the no of unique values of Postal codes, no duplicate values are present 

Check for any 'Not assigned ' values in the Neighborhood column:

In [8]:
filtered_df = Toronto_zipcodes[Toronto_zipcodes['Neighbourhood']=='Not assigned']

In [9]:
filtered_df.shape

(0, 3)

0 rows implies none of the Neighborhoods have 'Not assigned' value

As the no of samples are quite small(=103) ,  visually check the dataframe for any missing or null values

In [10]:
pd.set_option('display.max_rows',None)
Toronto_zipcodes

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Vic

None of the values are missing .

In [11]:
Toronto_zipcodes.shape

(103, 3)

# This conlcudes the first part of the project

The next part of the project involves finding the latitudes and longitudes of each zip code area and adding them as columns to the Toronto_zipcodes dataframe

In [12]:
Toronto_zipcodes.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

The related data is available at http://cocl.us/Geospatial_data . The csv file has been downloaded from here.

In [13]:
latlng_df = pd.read_csv('Geospatial_Coordinates.csv')

In [17]:
latlng_df

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

In [14]:
latlng_df.shape

(103, 3)

In [15]:
# define a function to get latitude and longitude from latlng_df using postal codes
def getLatLong(postal_code):
    """The function returns a numpy array of size (1,2) containing latitude and longitude 
    for a given postal code"""
    arr = latlng_df[latlng_df['Postal Code']==postal_code][['Latitude','Longitude']].values
    return arr    

In [16]:
#use the function getLatLong to obtain Latitude and Longitude information and add to the original dataframe
Toronto_zipcodes['Latitude'] = Toronto_zipcodes[['Postal Code']].applymap(lambda x : getLatLong(x)[0][0])
Toronto_zipcodes['Longitude'] = Toronto_zipcodes[['Postal Code']].applymap(lambda x : getLatLong(x)[0][1])

In [17]:
Toronto_zipcodes

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

# This concludes the second part of the project

Now we will use Foursquare API to query for upto 100 venues in each neighborhood(postal code) within a specified radius.


In [22]:
#input variables for Foursquare API
CLIENT_ID = 'PPH2EZN52CRO1PU4MASN0FFY1TOCS42GDRDN2INGULE0VRGQ'
CLIENT_SECRET = 'LWSF4UHMH4ETVS22DE4BLYOAN2PEDWVPQ3QJY4RW5KZIS5XO'
VERSION = 20200719
RADIUS=500# radius of 500m around the neighborhood
LIMIT =100 # limit results to 100 venues per day

In [23]:
import requests
from pandas.io.json import json_normalize

In [24]:
# first explore the venues near one particular neighborhood
# for Postal Code M3A
latitude = 43.725882
longitude = -79.315572

In [25]:
#use foursquare api
url = 'https://api.foursquare.com/v2/venues/explore?\
       &client_id={}\
       &client_secret={}\
       &v={}\
       &ll={},{}\
       &radius={}\
       &limit={}'.format(
       CLIENT_ID,CLIENT_SECRET,VERSION,latitude,longitude,RADIUS,LIMIT)

In [26]:
# Use requests to convert the response to json file
results = requests.get(url).json()


In [27]:
# Explore the results 
results

{'meta': {'code': 200, 'requestId': '5f17f4826b52fb1f37f86464'},
 'response': {'headerLocation': 'Bermondsey',
  'headerFullLocation': 'Bermondsey, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.7303820045, 'lng': -79.30935658242831},
   'sw': {'lat': 43.721381995499996, 'lng': -79.3217874175717}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4c633acb86b6be9a61268e34',
       'name': 'Victoria Village Arena',
       'location': {'lat': 43.72348055545508,
        'lng': -79.31563520925143,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.72348055545508,
          'lng': -79.31563520925143}],
        'distance': 267,
        'cc': 'CA',
        'country': 'Canada',
        'formatte

In [28]:
# items of interest are under response--groups--first item---items
#using json normalize to convert this into a dataframe
venues = json_normalize(results['response']['groups'][0]['items'])

In [29]:
# Find the category of the venues  and assign it to a new column
venues['venue.category'] = venues['venue.categories'].apply(lambda x : x[0]['name'])

In [30]:
pd.set_option('display.max_columns',None)
venues.head()

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   
2              0  [{'summary': 'This spot is popular', 'type': '...   
3              0  [{'summary': 'This spot is popular', 'type': '...   
4              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-4c633acb86b6be9a61268e34-0   
1  e-0-4f3ecce6e4b0587016b6f30d-1   
2  e-0-4bbe904a85fbb713420d7167-2   
3  e-0-4d689350b6f46dcb77ee15b2-3   
4  e-0-4b8fe578f964a520106833e3-4   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d185941735', 'name': 'H...   
1  [{'id': '4def73e84765ae376e57713a', 'name': 'P...   
2  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...   
3  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...   
4  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...   

                   venue.id     venue.location.address venue.location.cc  \
0  4c633acb86b6be9a61268e34                        NaN                CA   
1  4f3ecce6e4b0587016b6f30d  1733 Eglinton Avenue East                CA   
2  4bbe904a85fbb713420d7167     1733 Eglinton Ave East                CA   
3  4d689350b6f46dcb77ee15b2                        NaN                CA   
4  4b8fe578f964a520106833e3       1733 Eglinton Ave. E                CA   

  venue.location.city venue.location.country venue.location.crossStreet  \
0                 NaN                 Canada                        NaN   
1             Toronto                 Canada                 Bermondsey   
2             Toronto                 Canada           at Bermondsey Rd   
3                 NaN                 Canada                        NaN   
4             Toronto                 Canada          at Bermondsey Rd.   

   venue.location.distance                    venue.location.formattedAddress  \
0                      267                                           [Canada]   
1                      224  [1733 Eglinton Avenue East (Bermondsey), Toron...   
2                      202  [1733 Eglinton Ave East (at Bermondsey Rd), To...   
3                      197                                           [Canada]   
4                      218  [1733 Eglinton Ave. E (at Bermondsey Rd.), Tor...   

                       venue.location.labeledLatLngs  venue.location.lat  \
0  [{'label': 'display', 'lat': 43.72348055545508...           43.723481   
1  [{'label': 'display', 'lat': 43.72581876267242...           43.725819   
2  [{'label': 'display', 'lat': 43.72551663171475...           43.725517   
3  [{'label': 'display', 'lat': 43.72705130603407...           43.727051   
4  [{'label': 'display', 'lat': 43.72582372635219...           43.725824   

   venue.location.lng venue.location.postalCode venue.location.state  \
0          -79.315635                       NaN                  NaN   
1          -79.312785                       NaN                   ON   
2          -79.313103                   M4A 1J8                   ON   
3          -79.317418                       NaN                  NaN   
4          -79.312860                   M4A 1J8                   ON   

               venue.name  venue.photos.count venue.photos.groups  \
0  Victoria Village Arena                   0                  []   
1               Portugril                   0                  []   
2             Tim Hortons                   0                  []   
3                The Frig                   0                  []   
4              Pizza Nova                   0                  []   

          venue.category  
0           Hockey Arena  
1  Portuguese Restaurant  
2            Coffee Shop  
3      French Restaurant  
4            Pizza Place

In [31]:
#  redefine the venues df to contain only those features which are required for exploration,ie venue_name,category,latitude,longitude
venues = venues[['venue.name','venue.category','venue.location.lat','venue.location.lng']]

In [32]:
venues.head()

venue.name         venue.category  venue.location.lat  \
0  Victoria Village Arena           Hockey Arena           43.723481   
1               Portugril  Portuguese Restaurant           43.725819   
2             Tim Hortons            Coffee Shop           43.725517   
3                The Frig      French Restaurant           43.727051   
4              Pizza Nova            Pizza Place           43.725824   

   venue.location.lng  
0          -79.315635  
1          -79.312785  
2          -79.313103  
3          -79.317418  
4          -79.312860

In [33]:
# We will add columns for the postal code and neighborhood
venues.insert(0,'Postal Code',Toronto_zipcodes.iloc[0,0])

In [34]:
venues.insert(1,'Neighborhood',Toronto_zipcodes.iloc[0,2])

In [35]:
venues.head()

Postal Code Neighborhood              venue.name         venue.category  \
0         M3A    Parkwoods  Victoria Village Arena           Hockey Arena   
1         M3A    Parkwoods               Portugril  Portuguese Restaurant   
2         M3A    Parkwoods             Tim Hortons            Coffee Shop   
3         M3A    Parkwoods                The Frig      French Restaurant   
4         M3A    Parkwoods              Pizza Nova            Pizza Place   

   venue.location.lat  venue.location.lng  
0           43.723481          -79.315635  
1           43.725819          -79.312785  
2           43.725517          -79.313103  
3           43.727051          -79.317418  
4           43.725824          -79.312860

In [36]:
Toronto_zipcodes.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Filter the database to those Boroughs which contain 'Toronto' in their name

In [37]:
Toronto_zipcodes = Toronto_zipcodes[Toronto_zipcodes['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [38]:
Toronto_zipcodes.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [39]:
# define a function to get nearby venues to a neighborhood
def getNearbyVenues(row):
    
    lat = Toronto_zipcodes.loc[row,'Latitude']
    long = Toronto_zipcodes.loc[row,'Longitude']
    
    url = 'https://api.foursquare.com/v2/venues/explore?\
           &client_id={}\
           &client_secret={}\
           &v={}\
           &ll={},{}\
           &radius={}\
           &limit={}'.format(
           CLIENT_ID,
           CLIENT_SECRET,
           VERSION,
           lat,
           long,
           RADIUS,
           LIMIT)
    results = requests.get(url).json()
    n_nby_venues = json_normalize(results['response']['groups'][0]['items'])
    
    string = 'venue.categories'
    
    if string in n_nby_venues.columns:
        n_nby_venues['type'] = n_nby_venues['venue.categories'].apply(lambda x : x[0]['name'])
        n_nby_venues['venue_lat'] = n_nby_venues['venue.location.lat']
        n_nby_venues['venue_long'] = n_nby_venues['venue.location.lng']
        n_nby_venues = n_nby_venues[['venue.name','type','venue_lat','venue_long']]
        n_nby_venues['PostCode'] = Toronto_zipcodes.loc[row,'Postal Code']
        n_nby_venues['Neighbourhood'] = Toronto_zipcodes.loc[row,'Neighbourhood']
        n_nby_venues['Nborhood_latitude'] = Toronto_zipcodes.loc[row,'Latitude']
        n_nby_venues['Nborhood_longitude'] = Toronto_zipcodes.loc[row,'Longitude']
    
        return n_nby_venues
    else:
        return None
    

In [40]:
nby_venues = pd.DataFrame()
#loop through for each nborhood and find venues for them
for row in range(0,len(Toronto_zipcodes)):
    n_nby_venue = getNearbyVenues(row)
    nby_venues = pd.concat([nby_venues,n_nby_venue],ignore_index=True)
nby_venues.shape

(1644, 8)

In [41]:
nby_venues.tail(50)

venue.name                      type  venue_lat  \
1594                 The Yoga Sanctuary               Yoga Studio  43.661499   
1595                         Wow! Sushi          Sushi Restaurant  43.668514   
1596         The Betty Oliphant Theatre                   Theater  43.664315   
1597                    The Blake House       American Restaurant  43.664468   
1598                      Tokyo Kitchen       Japanese Restaurant  43.668783   
1599                         NC Salon +   Health & Beauty Service  43.669406   
1600                      Onnki Donburi       Japanese Restaurant  43.669757   
1601                  The Anndore House                     Hotel  43.668801   
1602                Ritz Caribbean Food      Caribbean Restaurant  43.669545   
1603            Cawthra Square Dog Park                   Dog Run  43.666583   
1604                    The Salad House  Mediterranean Restaurant  43.669508   
1605                         Rolltation          Sushi Restaurant  43.669388   
1606                        Constantine  Mediterranean Restaurant  43.668773   
1607                          Starbucks               Coffee Shop  43.662407   
1608                     Tika Tea House           Bubble Tea Shop  43.668797   
1609               Church Street Garage                   Gay Bar  43.664710   
1610                          Starbucks               Coffee Shop  43.668071   
1611                       Naan & Kabob         Afghan Restaurant  43.669005   
1612                          Starbucks               Coffee Shop  43.661527   
1613                            O. Noir                Restaurant  43.669145   
1614                     Kokoni Izakaya       Japanese Restaurant  43.664181   
1615                          Joe Fresh            Clothing Store  43.661956   
1616               Noah's Natural Foods         Food & Drink Shop  43.668532   
1617                    Town Inn Suites                     Hotel  43.669056   
1618                 Bishop and Belcher                       Pub  43.670096   
1619                          The Drink               Coffee Shop  43.664167   
1620                              Flash                Strip Club  43.664319   
1621              Alexander Wood Statue          Sculpture Garden  43.663702   
1622                     Croissant Tree               Coffee Shop  43.669575   
1623                        Asahi Sushi          Sushi Restaurant  43.669874   
1624                                A&W      Fast Food Restaurant  43.666415   
1625              Rorschach Brewing Co.                   Brewery  43.663483   
1626         Leslieville Farmers Market            Farmers Market  43.664901   
1627                       The Sidekick                Comic Shop  43.664484   
1628                        Chino Locos             Burrito Place  43.664653   
1629             Queen Margherita Pizza               Pizza Place  43.664685   
1630                        Chick-n-Joy      Fast Food Restaurant  43.665181   
1631           Ashbridges Bay Skatepark                Skate Park  43.662548   
1632                     The Green Wood                Restaurant  43.664728   
1633  East End Garden Centre & Hardware             Garden Center  43.664564   
1634             Amin Car Repair Garage             Auto Workshop  43.663544   
1635               The Ashbridge Estate                    Garden  43.664691   
1636               TTC Russell Division        Light Rail Station  43.664908   
1637            Jonathan Ashbridge Park                      Park  43.664702   
1638                       The Ten Spot                       Spa  43.664815   
1639                   Olliffe On Queen                   Butcher  43.664503   
1640                    TTC Stop #03049        Light Rail Station  43.664470   
1641          Greenwood Cigar & Variety                Smoke Shop  43.664538   
1642                        ONE Academy      Gym / Fitness Center  43.662253   
1643               Revolution

In [42]:
# groupby the neighborhoods in nby_venues
grpby_nbrhoods = nby_venues.groupby('Neighbourhood').count()

In [43]:
grpby_nbrhoods

venue.name  type  \
Neighbourhood                                                          
Berczy Park                                                 59    59   
Brockton, Parkdale Village, Exhibition Place                23    23   
Business reply mail Processing Centre, South Ce...          19    19   
CN Tower, King and Spadina, Railway Lands, Harb...          17    17   
Central Bay Street                                          65    65   
Christie                                                    17    17   
Church and Wellesley                                        76    76   
Commerce Court, Victoria Hotel                             100   100   
Davisville                                                  32    32   
Davisville North                                             9     9   
Dufferin, Dovercourt Village                                13    13   
First Canadian Place, Underground city                     100   100   
Forest Hill North & West, Forest Hill Road Park              5     5   
Garden District, Ryerson                                   100   100   
Harbourfront East, Union Station, Toronto Islands          100   100   
High Park, The Junction South                               26    26   
India Bazaar, The Beaches West                              21    21   
Kensington Market, Chinatown, Grange Park                   68    68   
Lawrence Park                                                4     4   
Little Portugal, Trinity                                    47    47   
Moore Park, Summerhill East                                  1     1   
North Toronto West, Lawrence Park                           20    20   
Parkdale, Roncesvalles                                      14    14   
Queen's Park, Ontario Provincial Government                 32    32   
Regent Park, Harbourfront                                   46    46   
Richmond, Adelaide, King                                   100   100   
Rosedale                                                     4     4   
Roselawn                                                     2     2   
Runnymede, Swansea                                          35    35   
St. James Town                                              85    85   
St. James Town, Cabbagetown                                 46    46   
Stn A PO Boxes                                              98    98   
Studio District                                             41    41   
Summerhill West, Rathnelly, South Hill, Forest ...          16    16   
The Annex, North Midtown, Yorkville                         21    21   
The Beaches                                                  5     5   
The Danforth West, Riverdale                                42    42   
Toronto Dominion Centre, Design Exchange                   100   100   
University of Toronto, Harbord                              35    35   

                                                    venue_lat  venue_long  \
Neighbourhood                                                               
Berczy Park                                                59          59   
Brockton, Parkdale Village, Exhibition Place               23          23   
Business reply mail Processing Centre, South Ce...         19          19   
CN Tower, King and Spadina, Railway Lands, Harb...         17          17   
Central Bay Street                                         65          65   
Christie                                                   17          17   
Church and Wellesley                                       76          76   
Commerce Court, Victoria Hotel                            100         100   
Davisville                                                 32          32   
Davisville North                                            9           9   
Dufferin, Dovercourt Village                               13          13   
First Canadian Place, Underground city                    100         100   
Forest Hill North & W

In [44]:
# find the value counts of venue types
nby_venues['type'].value_counts()

Coffee Shop                        158
Café                                90
Restaurant                          54
Hotel                               40
Italian Restaurant                  39
Park                                36
Japanese Restaurant                 32
Bakery                              30
Gym                                 29
Pizza Place                         29
Bar                                 28
Clothing Store                      23
Sandwich Place                      23
Seafood Restaurant                  23
Sushi Restaurant                    23
Pub                                 22
American Restaurant                 21
Thai Restaurant                     19
Vegetarian / Vegan Restaurant       18
Beer Bar                            18
Breakfast Spot                      17
Bookstore                           17
Gastropub                           16
Ice Cream Shop                      16
Diner                               16
Deli / Bodega            

In [45]:
#apply one-hot encoding to the venue types column
venue_type_dummies = pd.get_dummies(nby_venues['type'])

In [46]:
venue_type_dummies.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0

In [47]:
venue_type_dummies.shape

(1644, 233)

In [48]:
# add back to nby_venues dataframe
nby_venues = pd.concat([nby_venues,venue_type_dummies],axis=1)


In [49]:
nby_venues.head()

venue.name                 type  venue_lat  venue_long  \
0        Roselle Desserts               Bakery  43.653447  -79.362017   
1           Tandem Coffee          Coffee Shop  43.653559  -79.361809   
2  Cooper Koo Family YMCA  Distribution Center  43.653249  -79.358008   
3     Body Blitz Spa East                  Spa  43.654735  -79.359874   
4          Impact Kitchen           Restaurant  43.656369  -79.356980   

  PostCode              Neighbourhood  Nborhood_latitude  Nborhood_longitude  \
0      M5A  Regent Park, Harbourfront           43.65426          -79.360636   
1      M5A  Regent Park, Harbourfront           43.65426          -79.360636   
2      M5A  Regent Park, Harbourfront           43.65426          -79.360636   
3      M5A  Regent Park, Harbourfront           43.65426          -79.360636   
4      M5A  Regent Park, Harbourfront           43.65426          -79.360636   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0

 Remove the venue.name, type and venue.lat and long columns

In [51]:
nby_venues.drop(['venue.name','type','venue_lat','venue_long'],axis=1,inplace=True)

In [52]:
nby_venues.head(50)

PostCode                                Neighbourhood  Nborhood_latitude  \
0       M5A                    Regent Park, Harbourfront          43.654260   
1       M5A                    Regent Park, Harbourfront          43.654260   
2       M5A                    Regent Park, Harbourfront          43.654260   
3       M5A                    Regent Park, Harbourfront          43.654260   
4       M5A                    Regent Park, Harbourfront          43.654260   
5       M5A                    Regent Park, Harbourfront          43.654260   
6       M5A                    Regent Park, Harbourfront          43.654260   
7       M5A                    Regent Park, Harbourfront          43.654260   
8       M5A                    Regent Park, Harbourfront          43.654260   
9       M5A                    Regent Park, Harbourfront          43.654260   
10      M5A                    Regent Park, Harbourfront          43.654260   
11      M5A                    Regent Park, Harbourfront          43.654260   
12      M5A                    Regent Park, Harbourfront          43.654260   
13      M5A                    Regent Park, Harbourfront          43.654260   
14      M5A                    Regent Park, Harbourfront          43.654260   
15      M5A                    Regent Park, Harbourfront          43.654260   
16      M5A                    Regent Park, Harbourfront          43.654260   
17      M5A                    Regent Park, Harbourfront          43.654260   
18      M5A                    Regent Park, Harbourfront          43.654260   
19      M5A                    Regent Park, Harbourfront          43.654260   
20      M5A                    Regent Park, Harbourfront          43.654260   
21      M5A                    Regent Park, Harbourfront          43.654260   
22      M5A                    Regent Park, Harbourfront          43.654260   
23      M5A                    Regent Park, Harbourfront          43.654260   
24      M5A                    Regent Park, Harbourfront          43.654260   
25      M5A                    Regent Park, Harbourfront          43.654260   
26      M5A                    Regent Park, Harbourfront          43.654260   
27      M5A                    Regent Park, Harbourfront          43.654260   
28      M5A                    Regent Park, Harbourfront          43.654260   
29      M5A                    Regent Park, Harbourfront          43.654260   
30      M5A                    Regent Park, Harbourfront          43.654260   
31      M5A                    Regent Park, Harbourfront          43.654260   
32      M5A                    Regent Park, Harbourfront          43.654260   
33      M5A                    Regent Park, Harbourfront          43.654260   
34      M5A                    Regent Park, Harbourfront          43.654260   
35      M5A                    Regent Park, Harbourfront          43.654260   
36      M5A                    Regent Park, Harbourfront          43.654260   
37      M5A                    Regent Park, Harbourfront          43.654260   
38      M5A                    Regent Park, Harbourfront          43.654260   
39      M5A                    Regent Park, Harbourfront          43.654260   
40      M5A                    Regent Park, Harbourfront          43.654260   
41      M5A                    Regent Park, Harbourfront          43.654260   
42      M5A                    Regent Park, Harbourfront          43.654260   
43      M5A                    Regent Park, Harbourfront          43.654260   
44      M5A                    Regent Park, Harbourfront          43.654260   
45      M5A                    Regent Park, Harbourfront          43.654260   
46      M7A  Queen's Park, Ontario Provincial Government          43.662301   
47      M7A  Queen's Park, Ontario Provincial Government          43.662301   
48      M7A  Queen's Park, Ontario Provincial Government          43.662301   
49      M7A  Queen's Park, Ontario Provincial Governm

In [53]:
nby_venues.shape

(1644, 237)

In [56]:
nby_venues_grpd = nby_venues.groupby('Neighbourhood').mean()

In [57]:
nby_venues_grpd

Nborhood_latitude  \
Neighbourhood                                                           
Berczy Park                                                 43.644771   
Brockton, Parkdale Village, Exhibition Place                43.636847   
Business reply mail Processing Centre, South Ce...          43.662744   
CN Tower, King and Spadina, Railway Lands, Harb...          43.628947   
Central Bay Street                                          43.657952   
Christie                                                    43.669542   
Church and Wellesley                                        43.665860   
Commerce Court, Victoria Hotel                              43.648199   
Davisville                                                  43.704324   
Davisville North                                            43.712751   
Dufferin, Dovercourt Village                                43.669005   
First Canadian Place, Underground city                      43.648429   
Forest Hill North & West, Forest Hill Road Park             43.696948   
Garden District, Ryerson                                    43.657162   
Harbourfront East, Union Station, Toronto Islands           43.640816   
High Park, The Junction South                               43.661608   
India Bazaar, The Beaches West                              43.668999   
Kensington Market, Chinatown, Grange Park                   43.653206   
Lawrence Park                                               43.728020   
Little Portugal, Trinity                                    43.647927   
Moore Park, Summerhill East                                 43.689574   
North Toronto West, Lawrence Park                           43.715383   
Parkdale, Roncesvalles                                      43.648960   
Queen's Park, Ontario Provincial Government                 43.662302   
Regent Park, Harbourfront                                   43.654260   
Richmond, Adelaide, King                                    43.650571   
Rosedale                                                    43.679563   
Roselawn                                                    43.711695   
Runnymede, Swansea                                          43.651571   
St. James Town                                              43.651494   
St. James Town, Cabbagetown                                 43.667967   
Stn A PO Boxes                                              43.646435   
Studio District                                             43.659525   
Summerhill West, Rathnelly, South Hill, Forest ...          43.686412   
The Annex, North Midtown, Yorkville                         43.672710   
The Beaches                                                 43.676357   
The Danforth West, Riverdale                                43.679557   
Toronto Dominion Centre, Design Exchange                    43.647177   
University of Toronto, Harbord                              43.662696   

                                                    Nborhood_longitude  \
Neighbourhood                                                            
Berczy Park                                                 -79.373306   
Brockton, Parkdale Village, Exhibition Place                -79.428191   
Business reply mail Processing Centre, South Ce...          -79.321558   
CN Tower, King and Spadina, Railway Lands, Harb...          -79.394420   
Central Bay Street                                          -79.387383   
Christie                                                    -79.422564   
Church and Wellesley                                        -79.383160   
Commerce Court, Victoria Hotel                              -79.379817   
Davisville                                                  -79.388790   
Davisville North                                            -79.390197   
Dufferin, Dovercourt Village                                -79.442259   
First Canadian Place, Underground city                      -79.382280   
Forest Hill North & We

In [58]:
nby_venues_grpd.shape

(39, 235)

# KMeans clustering

 We will group the nbrhoods into similar clusters based on the venues in the neighborhood

In [59]:
from sklearn.cluster import KMeans

In [60]:
# define our input features
X = nby_venues_grpd.drop(['Nborhood_latitude','Nborhood_longitude'],axis=1)

In [61]:
X.head()

Afghan Restaurant  \
Neighbourhood                                                           
Berczy Park                                                       0.0   
Brockton, Parkdale Village, Exhibition Place                      0.0   
Business reply mail Processing Centre, South Ce...                0.0   
CN Tower, King and Spadina, Railway Lands, Harb...                0.0   
Central Bay Street                                                0.0   

                                                     Airport  \
Neighbourhood                                                  
Berczy Park                                         0.000000   
Brockton, Parkdale Village, Exhibition Place        0.000000   
Business reply mail Processing Centre, South Ce...  0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...  0.058824   
Central Bay Street                                  0.000000   

                                                    Airport Food Court  \
Neighbourhood                                                            
Berczy Park                                                   0.000000   
Brockton, Parkdale Village, Exhibition Place                  0.000000   
Business reply mail Processing Centre, South Ce...            0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...            0.058824   
Central Bay Street                                            0.000000   

                                                    Airport Gate  \
Neighbourhood                                                      
Berczy Park                                             0.000000   
Brockton, Parkdale Village, Exhibition Place            0.000000   
Business reply mail Processing Centre, South Ce...      0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...      0.058824   
Central Bay Street                                      0.000000   

                                                    Airport Lounge  \
Neighbourhood                                                        
Berczy Park                                               0.000000   
Brockton, Parkdale Village, Exhibition Place              0.000000   
Business reply mail Processing Centre, South Ce...        0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...        0.117647   
Central Bay Street                                        0.000000   

                                                    Airport Service  \
Neighbourhood                                                         
Berczy Park                                                0.000000   
Brockton, Parkdale Village, Exhibition Place               0.000000   
Business reply mail Processing Centre, South Ce...         0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...         0.117647   
Central Bay Street                                         0.000000   

                                                    Airport Terminal  \
Neighbourhood                                                          
Berczy Park                                                 0.000000   
Brockton, Parkdale Village, Exhibition Place                0.000000   
Business reply mail Processing Centre, South Ce...          0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...          0.117647   
Central Bay Street                                          0.000000   

                                                    American Restaurant  \
Neighbourhood                                                             
Berczy Park                                                         0.0   
Brockton, Parkdale Village, Exhibition Place                        0.0   
Business reply mail Processing Centre, South Ce...                  0.0   
CN Tower, King and Spadina, Railway Lands, Harb...                  0.0   
Central Bay Street                                                  0.0   

                                                    Antique Shop  Aquarium 

In [62]:
k = 3 # no. of labels
kmean = KMeans(n_clusters=k,random_state=1)

In [63]:
kmean.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=3, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=1, tol=0.0001, verbose=0)

In [64]:
labels = kmean.labels_

In [65]:
labels

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 2, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [66]:
#insert the label column in the nby_venues_grpd database
nby_venues_grpd.insert(0,'labels',value=labels)

In [67]:
nby_venues_grpd.head()

labels  Nborhood_latitude  \
Neighbourhood                                                                   
Berczy Park                                              0          43.644771   
Brockton, Parkdale Village, Exhibition Place             0          43.636847   
Business reply mail Processing Centre, South Ce...       0          43.662744   
CN Tower, King and Spadina, Railway Lands, Harb...       0          43.628947   
Central Bay Street                                       0          43.657952   

                                                    Nborhood_longitude  \
Neighbourhood                                                            
Berczy Park                                                 -79.373306   
Brockton, Parkdale Village, Exhibition Place                -79.428191   
Business reply mail Processing Centre, South Ce...          -79.321558   
CN Tower, King and Spadina, Railway Lands, Harb...          -79.394420   
Central Bay Street                                          -79.387383   

                                                    Afghan Restaurant  \
Neighbourhood                                                           
Berczy Park                                                       0.0   
Brockton, Parkdale Village, Exhibition Place                      0.0   
Business reply mail Processing Centre, South Ce...                0.0   
CN Tower, King and Spadina, Railway Lands, Harb...                0.0   
Central Bay Street                                                0.0   

                                                     Airport  \
Neighbourhood                                                  
Berczy Park                                         0.000000   
Brockton, Parkdale Village, Exhibition Place        0.000000   
Business reply mail Processing Centre, South Ce...  0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...  0.058824   
Central Bay Street                                  0.000000   

                                                    Airport Food Court  \
Neighbourhood                                                            
Berczy Park                                                   0.000000   
Brockton, Parkdale Village, Exhibition Place                  0.000000   
Business reply mail Processing Centre, South Ce...            0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...            0.058824   
Central Bay Street                                            0.000000   

                                                    Airport Gate  \
Neighbourhood                                                      
Berczy Park                                             0.000000   
Brockton, Parkdale Village, Exhibition Place            0.000000   
Business reply mail Processing Centre, South Ce...      0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...      0.058824   
Central Bay Street                                      0.000000   

                                                    Airport Lounge  \
Neighbourhood                                                        
Berczy Park                                               0.000000   
Brockton, Parkdale Village, Exhibition Place              0.000000   
Business reply mail Processing Centre, South Ce...        0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...        0.117647   
Central Bay Street                                        0.000000   

                                                    Airport Service  \
Neighbourhood                                                         
Berczy Park                                                0.000000   
Brockton, Parkdale Village, Exhibition Place               0.000000   
Business reply mail Processing Centre, South Ce...         0.000000   
CN Tower, King and Spadina, Railway Lands, Harb...         0.117647   
Central Bay Street                                         0.000000   

                   

# Plotting

 Make a map of Toronto using Folium and put the markers for nborhoods based on their labels

In [68]:
import folium
from geopy import Nominatim

In [69]:
# find the latitude and longitude for Toronto using Geopy
address = 'Toronto,Canada'
geolocator = Nominatim(user_agent='Toronto_explorer')
location = geolocator.geocode(address)
T_latitude = location.latitude
T_longitude = location.longitude

In [70]:
T_latitude

43.6534817

In [71]:
T_longitude

-79.3839347

In [72]:
# draw map of Toronto using above coordinates
Toronto_map = folium.Map(location=[T_latitude,T_longitude],zoom_start=12)

In [73]:
Toronto_map

In [74]:
# place the markers for neighborhoods in the map based on the labels
colors = ['Red','Blue','Green']
#loop through for each label
for i in range(0,3):
    venues_by_label = nby_venues_grpd[nby_venues_grpd['labels']==i].iloc[:,0:3]
    #loop through each nborhood in the particular label
    for lat,long,nbrhood in zip(
        venues_by_label['Nborhood_latitude'],
        venues_by_label['Nborhood_longitude'],
        venues_by_label.index):
        
            label = "{}".format(nbrhood)

            folium.CircleMarker(
            [lat,long],
            radius=3,
            popup=label,
            color=colors[i],
            parse_html=True).add_to(Toronto_map)



In [75]:
# Map of Toronto with neighborhoods grouped according to nearby venues
Toronto_map

This concludes the project. From the map we can see that most of the neighborhoods are similar based on the type of venues they have in the neighborhood